### PR TITLE
Add GIF support

### PR DESCRIPTION
**Pull request by bontibon**
- Only the chat log (`LogTextBrowser`) supports them in this patch
- GIFs only play when they are hovered over

Updates #1698

Edit: Right now, "Save Image As..." only saves the current frame.

### CODE REVIEW

**Comment by Natenom** [comment]
* They start playing when hovered but then the do never again stop. Could one add a "Stop all animations" to their context menu?
* When the imagemessagelength is set to lets say 300KiB then one single image is no problem, but an animated gif with the same size lets the client freeze for some time. For example this: https://en.wikipedia.org/wiki/File:Newtons_cradle_animation_book_2.gif - Mumble consumes 100% on my i5 M 520 (2.40GHz) while playling the gif. Sometimes playback becomes very laggy when changing the focus between chaglog and channel tree. Maybe one could add a animatedimagemessagelength for that case?

**Comment by mkrautz** [comment]
Looks good functionality-wise.

But performance-wise, I think it's too heavy to ship as-is.

Do we know what is expensive here? Is it because it's a data URL?

Is it invalidating the log view with the event?

etc.

[But I would like it if we could limit it to only play the gifs when the mouse if hovered over it]

File: src/mumble/CustomElements.cpp [hunk 1]
```diff
@@ -69,6 +69,57 @@ void LogTextBrowser::scrollLogToBottom() {
 	verticalScrollBar()->setValue(verticalScrollBar()->maximum());
 }
 
+void LogTextBrowser::mouseMoveEvent(QMouseEvent *e) {
+	do {
+		QTextCursor cursor = cursorForPosition(e->pos());
+		QTextCharFormat fmt = cursor.charFormat();
+		if (fmt.objectType() == QTextFormat::NoObject) {
+			cursor.movePosition(QTextCursor::NextCharacter);
+			fmt = cursor.charFormat();
```
Comment (Kissaki) [inline src/mumble/CustomElements.cpp]: What is this case for?

**Comment by Kissaki** [comment]
> But performance-wise, I think it's too heavy to ship as-is.

@mkrautz How did you determine that? Just a guess from the code? Or did you experience performance issues? "Do we know what is expensive here?" implies that you did?
At least for the small file sizes I am allowed on my server, I had no issues.

File: src/mumble/CustomElements.cpp (hunk 1 above)
Comment (bontibon) [inline src/mumble/CustomElements.cpp]: Taken from https://github.com/mumble-voip/mumble/blob/master/src/mumble/MainWindow.cpp#L718 (which in turn was probably taken from http://stackoverflow.com/a/18901483).

**Comment by Kissaki** [comment]
I simplified the code a bit, and looked into improving performance/selective update/redraw. I will push and create a new PR.

Issue I noticed: The MouseMoveEvents only trigger with the mouse inside the log area. When you move out to the left or right, the animation never stops playing.
The animation also plays when not directly over the image, but to the right of it if no character follows.
Not sure if this was the case before my changes, but the animation only plays with the cursor in the right half of the image, not on the left half.

**Comment by mkrautz** [comment]
It was Natenom comment (that big images cause the client to freeze) and also the fact that the PR is using a FontChange affecting the whole QTextBrowser to update a single message's shown frame.

I don't recall if I *saw* any perforamance issues myself.

File: src/mumble/CustomElements.cpp (hunk 1 above)
Comment (Kissaki) [inline src/mumble/CustomElements.cpp]: That’s also committed by you.  was able to reproduce the issue and solution. I'll document it.

**Comment by Kissaki** [comment]
I pushed to https://github.com/Kissaki/mumble/tree/pr-gif

- [ ] Saving a gif image from log does not work
- [ ] Image continues to play when moving outside of the log view to the left or right if there is no character/element before/after the image (we have no mouseMoveEvent to handle)
- [ ] We relayout the entire log content on every frame (FontChangeEvent)
- [ ] The animation does not play in the send message dialogue
- [ ] The same image added to the log will play on all instances when one of them is hovered

The log view extends [`QTextBrowser`](http://doc.qt.io/qt-5/qtextbrowser.html), which uses a [`QTextDocument`](http://doc.qt.io/Qt-5/qtextdocument.html). This is for [rich text](http://doc.qt.io/qt-5/richtext.html), including hyperlinks and (unanimated) images. Images are handled through a special replacement character, and a resource cache, in which the image data is stored and then displayed in place of the placeholder.

As there is no native support for animated images, we use QMovie to play the animation and put every frame into the resource buffer, replacing the image, and we then trigger a redraw through the `QEvent::FontChange` event (complete redraw).

The complete relayout/redraw is a big performance hit for just displaying an image - probably unacceptable.

**Comment by Kissaki** [comment]
Closing in favor of #2787

**Pull request closed without merging**